lib/aws/request.js

```diff
@@ -0,0 +1,234 @@
+'use strict';
+
+const _ = require('lodash');
+const memoize = require('memoizee');
+const PromiseQueue = require('promise-queue');
+const sdk = require('aws-sdk');
+const chalk = require('chalk');
+const ServerlessError = require('../../lib/serverless-error');
+const log = require('@serverless/utils/log');
+const HttpsProxyAgent = require('https-proxy-agent');
+const https = require('https');
+const fs = require('fs');
+const deepSortObjectByKey = require('../../lib/utils/deepSortObjectByKey');
+
+// Activate AWS SDK logging
+if (process.env.SLS_DEBUG) {
+  sdk.config.logger = log;
+}
+
+// Use HTTPS Proxy (Optional)
+const proxy =
+  process.env.proxy ||
+  process.env.HTTP_PROXY ||
+  process.env.http_proxy ||
+  process.env.HTTPS_PROXY ||
+  process.env.https_proxy;
+
+const proxyOptions = {};
+if (proxy) {
+  Object.assign(proxyOptions, new URL(proxy));
+}
+
+const ca = process.env.ca || process.env.HTTPS_CA || process.env.https_ca;
+
+let caCerts = [];
+
+if (ca) {
+  // Can be a single certificate or multiple, comma separated.
+  const caArr = ca.split(',');
+  // Replace the newline -- https://stackoverflow.com/questions/30400341
+  caCerts = caCerts.concat(caArr.map((cert) => cert.replace(/\\n/g, '\n')));
+}
+
+const cafile = process.env.cafile || process.env.HTTPS_CAFILE || process.env.https_cafile;
+
+if (cafile) {
+  // Can be a single certificate file path or multiple paths, comma separated.
+  const caPathArr = cafile.split(',');
+  caCerts = caCerts.concat(caPathArr.map((cafilePath) => fs.readFileSync(cafilePath.trim())));
+}
+
+if (caCerts.length > 0) {
+  Object.assign(proxyOptions, {
+    rejectUnauthorized: true,
+    ca: caCerts,
+  });
+}
+
+// Passes also certifications
+if (proxy) {
+  sdk.config.httpOptions.agent = new HttpsProxyAgent(proxyOptions);
+} else if (proxyOptions.ca) {
+  // Update the agent -- http://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/node-registering-certs.html
+  sdk.config.httpOptions.agent = new https.Agent(proxyOptions);
+}
+
+// Configure the AWS Client timeout (Optional).  The default is 120000 (2 minutes)
+const timeout = process.env.AWS_CLIENT_TIMEOUT || process.env.aws_client_timeout;
+if (timeout) {
+  sdk.config.httpOptions.timeout = parseInt(timeout, 10);
+}
+PromiseQueue.configure(Promise);
+const requestQueue = new PromiseQueue(2, Infinity);
+
+const MAX_RETRIES = (() => {
+  const userValue = Number(process.env.SLS_AWS_REQUEST_MAX_RETRIES);
+  return userValue >= 0 ? userValue : 4;
+})();
+
+const getServiceInstance = memoize(
+  (service, method) => {
+    // Checks regarding expectations on service object
+    if (
+      !_.isObject(service) ||
+      !_.isObject(service.params) ||
+      !_.isString(service.name) ||
+      !_.isObject(service.params.credentials)
```

Comment:
Let's validate this with [type](https://github.com/medikoo/type#readme) library. We already do it in some places, e.g. https://github.com/serverless/serverless/blob/b81fd0557f53e11597fe67bb138c09675d61e0c5/lib/Serverless.js#L33-L44

Reply:
It should definitely make the error message more meaningful.